Models,Migrations

mana-oceanpig opened the pull request.

mana-oceanpig committed:
Migrated and edited -db

mana-oceanpig committed:
Created -Models

ne-yamasaki commented on database/migrations/2014_10_12_000000_create_users_table.php:
```diff
@@ -18,15 +15,14 @@ public function up(): void
             $table->timestamp('email_verified_at')->nullable();
             $table->string('password');
             $table->rememberToken();
+            $table->timestamp('registered_at')->useCurrent(); // ユーザー登録日
```
「ユーザーテーブルのレコードが作成される日時」と「会員登録がなされる日時」を別に扱う必要があるなら、`registered_at` というカラムを設ける必要があるけど、あまりそのようなケースは経験がないです。

各日時が一致しているなら、`$table->timestamps()` で生成されるカラムの一つの `created_at` をユーザー登録日としてみなしてよいのでは？

ne-yamasaki commented on database/migrations/2024_06_24_062634_create_user_table.php:
```diff
@@ -0,0 +1,27 @@
+     */
+    public function up(): void
+    {
+        Schema::create('user', function (Blueprint $table) {
```
フレームワークのデフォルトの命名ルールは、テーブル名は複数形、モデルは単数形が基本なので、`users` というテーブル名の方が、Laravelを使う上で何かとラクですよ（デフォルトの命名ルールに反するテーブル名をつけることも可能ですが、その場合は、例えば `User`モデルにて利用するテーブル名は `user` ですよって指定する必要があります。指定がなければ、フレームワーク側が勝手に`User`というモデル名から、`users`ってテーブルが対応していると考えてくれます）

ne-yamasaki replied:
いや、`users`テーブルは一つ前のmigrationで定義しているのか。`user`テーブルは何のために存在している？

ne-yamasaki commented on database/migrations/2024_06_24_064914_create_conversation_messages_table.php:
```diff
@@ -0,0 +1,25 @@
+            $table->foreignId('conversation_id')->constrained()->onDelete('cascade'); // 会話IDへの外部キー
+            $table->text('message'); // メッセージ本文
+            $table->boolean('is_hidden')->default(false); // メッセージが隠されているかどうか
+            $table->TinyInt('role_id');
```
詳しくしらないのだけど、他の `Blurprint $table` のメソッド名が小文字で始まるのに、ここだけ `TinyInt()` となるの合ってますか？ `tinyInt()` な気がしまして

ne-yamasaki commented on app/Models/Conversation.php:
```diff
@@ -0,0 +1,25 @@
+    use HasFactory;
+
+    protected $fillable = [
+        'user_id', 'external_id', 'started_at', 'ended_at',
```
`migration`ファイルには、`started_at` と `ended_at` が定義されていないですよ？

ne-yamasaki commented on app/Models/Report.php:
```diff
@@ -0,0 +1,20 @@
+    use HasFactory;
+
+    protected $fillable = [
+        'user_id', 'summary', 'serotonin', 'dopamine', 'oxytocin', 'recommendations',
```
`pdf_path` が抜けてるかと

ne-yamasaki commented on app/Models/User.php:
```diff
@@ -2,44 +2,55 @@
-        'email',
-        'password',
+        'name', 'email', 'password',
     ];
```
必要ならば `registered_at` 、`last_login_at` を 追加する必要あり

ne-yamasaki commented on app/Models/User.php:
```diff
@@ -2,44 +2,55 @@
+    }
+
+    // アクセサ：登録日を取得
+    public function getRegisteredAtAttribute($value)
```
これはいまいち何をしたいかよく分からないです。
`$user->registered_at` にアクセスしたらDateTime型の 登録日時が取得できるので、モデルでは この `DateTime`オブジェクトを返せればよくて、`DateTime`オブジェクトを何かしらの文字列にフォーマット化するのは、Controller 側でやれば良いかと。そうしたら、場面に応じて、フォーマット化のルールも柔軟に変えられるので。

ne-yamasaki commented:
一応、Slackで伝えた内容を転載。

------

Githubのブランチ切っての開発とプルリクエスト、Good Tryです
おかげでフィードバックしやすくなりました。コメント書いたので確認くださいー！
https://github.com/mana-oceanpig/laravel_pj/pull/1/files

あと、フィードバック後の対応として、migrationファイルを修正する必要がある場合、以下の順序でお願いします。

1. 修正前に 今のスキーマをロールバックしておく (※1)
2. migration ファイルを含めたフィードバックを反映するコード修正をする
3. migration を再実行する

※1
Laravelの migration の実行管理は内部のデータベースで行われています。具体的には、各 migration file 名が記録されているので、ある migration file を実行した後、その migration file を編集しても、`php artisan migrate` を実行したところで、編集済の migration file は実行されません。

一度、実行した migration をやっぱ巻き戻すことをロールバックと言います。`php artisan migrate:rollback` とすれば最新のマイグレーションが巻き戻ります。もちろん、テーブル作成のマイグレーションだった場合、ロールバックすると登録されていたレコードも消えますし、テーブル自体が消えるので注意してください。

あと、`php artisan migrate` でマイグレーションを実行して、`php artisan migrate:rollback` でロールバックして、ということを繰り返すことはちょいちょいあるんですが、これがちゃんとできるようにするには、マイグレーションファイルの `up()` と `down()` の対をちゃんと整合性をあわせる必要があります。`down()` の手続きに漏れがあったりすると、エラーが生じる可能性があります。

mana-oceanpig committed:
fixed-migration

mana-oceanpig committed:
Edited-Models

・Conversation.php
→started_at / ended_atを削除
・Report.php
→pdf_pathを追加
・User.php
→created_at 、last_login_atをdatetimeに

mana-oceanpig committed:
Add files via upload

created -ConversationController

mana-oceanpig committed:
edited -ConversationController

mana-oceanpig committed:
edited -Models

lastMessageを追加
summarized_textをDBに実装したのでこちらをアップデート

mana-oceanpig committed:
added -SummarizedText

mana-oceanpig committed:
updated -Controllers

ne-yamasaki commented on database/migrations/2024_06_24_064914_create_conversation_messages_table.php:
```diff
@@ -0,0 +1,26 @@
+            $table->id();
+            $table->foreignId('conversation_id')->constrained()->onDelete('cascade'); // 会話IDへの外部キー
+            $table->text('message'); // メッセージ本文
+            $table->text('summarized_text')->nullable();
```
`summary`というカラム名に統合した方が良さそう

ne-yamasaki commented on database/migrations/2024_06_24_064914_create_conversation_messages_table.php:
```diff
@@ -0,0 +1,26 @@
+            $table->text('message'); // メッセージ本文
+            $table->text('summarized_text')->nullable();
+            $table->boolean('is_hidden')->default(false); // メッセージが隠されているかどうか
+            $table->tinyInteger('role_id');
```
絶対値が入っているなら `notNullable` にしておいた方が良さそう

ne-yamasaki commented on database/migrations/2024_06_24_071430_create_reports_table.php:
```diff
@@ -0,0 +1,28 @@
+{
+    public function up()
+    {
+        Schema::create('reports', function (Blueprint $table) {
```
`conversation_reports` というテーブル名に変更しましょう

ne-yamasaki replied:
`$table->foreignId('conversation_id')->constrained()->onDelete('cascade'); // Conversation IDへの外部キー` を追加ください

ne-yamasaki commented on app/Models/Report.php:
```diff
@@ -0,0 +1,20 @@
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+
+class Report extends Model
```
ここのモデル名を `ConversationReport` というモデル名に変更しましょう

mana-oceanpig committed:
edited -ConversationReport

ne-yamasaki commented on app/Http/Controllers/OpenAIController.php:
```diff
@@ -0,0 +1,15 @@
+
+use Illuminate\Http\Request;
+
+class OpenAIController extends Controller
```
これはなくても大丈夫！

ne-yamasaki commented on app/Http/Controllers/ConversationController.php:
```diff
@@ -0,0 +1,85 @@
+use Illuminate\Support\Facades\Storage;
+use Illuminate\Support\Facades\Log; 
+
+class ConversationController extends Controller
```
フレームワークの命名ルールに従うと、`ConversationsController`のような複数形にするのが一般的。
他、`ConversationMessagesController`, `ConversationReportsController` としておく方が、フレームワークを使う上で何かとラク出来たりする（今のままでも平気だけど、場合によっては暗黙のルールから外れているので、明示的に値を指定しないといけないとか面倒なことがあるので、ルールに乗っかっておくほうが無難です）

ne-yamasaki commented on database/migrations/2024_06_24_064914_create_conversation_messages_table.php:
```diff
@@ -0,0 +1,26 @@
+            $table->text('summary')->nullable();
+            $table->boolean('is_hidden')->default(false); // メッセージが隠されているかどうか
+            $table->tinyInteger('role_id')->notNullable();
+            $table->timestamps(); // created_at と updated_at を自動追加
```
`conversation_messages` にも、`$table->string('external_id')->nullable(); // 外部接続用のID` を追加しておきましょう

mana-oceanpig committed:
edited -name

mana-oceanpig committed:
added -Services
